### PR TITLE
Remove metrics

### PR DESCRIPTION
**Pull request by BartoszGrajdek**
### Details

During the TS migration of metrics, we had a [discussion on Slack](https://expensify.slack.com/archives/C01GTK53T8Q/p1708940842987279) if this feature is being used. No one raised any concerns so this PR removes metrics from the codebase (aside from PerformanceUtils which are used in a few other places).

### Related Issues

https://github.com/Expensify/App/issues/34341

### Manual Tests

Aside from making sure that onyx works and Performance utils didn't change in functionality there's nothing to test.

To make testing easier just run `npm run build` and paste dist file into expensify project.
1. Open Onyx.js inside node_modules/react-native-onyx
2. Find function init( and after the first conditional paste code below

```
    PerformanceUtils.setShouldDebugSetState(true);
    PerformanceUtils.logSetStateCall({ displayName: 'test', key: 'testKey' }, {}, {});
    PerformanceUtils.logSetStateCall({ displayName: 'test', key: 'testKey' }, { test: 1 }, { test2: 123 });
    PerformanceUtils.logSetStateCall({ displayName: 'test', key: 'testKey' }, { test: 1 }, { test: 123 });
    PerformanceUtils.logSetStateCall({ displayName: 'test', key: 'testKey' }, { test: 1, test2: 1 }, { test2: 123 });
    PerformanceUtils.logSetStateCall({ displayName: 'test', key: 'testKey' }, { test: 1, test2: 1, object: { test: 1, test2: 1 } }, { test2: 123, object: { test: 41, test2: 21 } });
```
3. You can play around with values, check what's printed in the console and compare them with how it used to work before

### Author Checklist

- [x] I linked the correct issue in the `### Related Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos

N/A

### CODE REVIEW

File: lib/PerformanceUtils.ts [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    debugSetState = debug;
+}
+
+/**
+ * Deep diff between two objects. Useful for figuring out what changed about an object from one render to the next so
+ * that state and props updates can be optimized.
+ */
+function diffObject<TObject extends UnknownObject, TBase extends UnknownObject>(object: TObject, base: TBase): UnknownObject {
+    return lodashTransform(object, (result: UnknownObject, value, key) => {
+        if (deepEqual(value, base[key])) {
```
Comment (blazejkustra) [inline lib/PerformanceUtils.ts:30]: You changed the method here, please make sure it works correctly

File: lib/PerformanceUtils.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+            return;
+        }
+
+        if (typeof value === 'object' && typeof base[key] === 'object') {
```
Comment (blazejkustra) [inline lib/PerformanceUtils.ts:34]: I was worried that `typeof value === 'object'` might work differently here, but I think it's fine 👍

**Review by blazejkustra — APPROVED** [review]
Let's test new `lib/PerformanceUtils.ts`, other than that LGTM!

File: lib/Onyx.js [hunk 3]
```diff
@@ -1554,21 +1554,7 @@ function setMemoryOnlyKeys(keyList) {
...
-    shouldSyncMultipleInstances = Boolean(global.localStorage),
-    debugSetState = false,
-} = {}) {
-    if (captureMetrics) {
-        // The code here is only bundled and applied when the captureMetrics is set
-        // eslint-disable-next-line no-use-before-define
-        applyDecorators();
-    }
-
+function init({keys = {}, initialKeyStates = {}, safeEvictionKeys = [], maxCachedKeysCount = 1000, shouldSyncMultipleInstances = Boolean(global.localStorage), debugSetState = false} = {}) {
```
Comment (kubabutkiewicz) [inline lib/Onyx.js:1557]: nab: shouldn't it be formatted like earlier? I think it was more readable

**Review by kubabutkiewicz — APPROVED** [review]

File: lib/Onyx.js (hunk 3 above)
Comment (BartoszGrajdek) [reply lib/Onyx.js:1557]: That's how prettier formatted it after removing one parameter 😅

File: lib/Onyx.js
Comment (fabioh8010) [inline lib/Onyx.js:1]: Please make sure the changes here are reflected in `Onyx.d.ts` too!

**Review by pecanoro — APPROVED** [review]

**Pull request merged**